### PR TITLE
Update bundle 042020

### PR DESCRIPTION
**Pull request by kheffah**
## [Commit 1](https://github.com/DigitalSlideArchive/HistomicsTK/pull/842/commits/16d7ec86dbc8a9049ae802749489fa4a3efb1506): Update + refactor girder utilities and workflows
1- Add documentation for using girder API,
2- Add a few girder convenience utilities,
3- Update workflows.
4- Make sure girder_utils test and workflows tests follow new standard
5- Make sure to update and document local DSA docker initialization for tests that require an authenticated girder client.
6- Standardize naming for tests so they are correctly discovered by pytest and tox.

## [Commit 2](https://github.com/DigitalSlideArchive/HistomicsTK/pull/842/commits/c44b8011c1bdc71ac8cdd37261bbdebbff1f1214):  Update + refactor annotations and masks
1- Update annotation and mask modules and utils
  - **IMPORTANT!**  This reverts + updates changes that were mistakenly merged into master in 
  annotations_to_masks_handler in #833)

2- Add annotations to object segmentation mask functionality (see #816)
3- Cherry pick from latest merges into master to avoid conflicts while working on this
4- Move test files for annotations and masks to standard location
5- Update annotation and masts tests (including polygon merger) to follow new standard and re-enable.

## [Commit 3](https://github.com/DigitalSlideArchive/HistomicsTK/pull/842/commits/3fe99f2fe82aaf527fb9e6554d9bb83fc8d0eba2): Local mongo backup and sqlite parser
1- Add functionality for mongo local backup and SQLite parser (see #803)
2- Add sqlalchemy to setup.py

## [Commit 4](https://github.com/DigitalSlideArchive/HistomicsTK/pull/842/commits/06a821a601b1aa0e4c9ca44a4d3bc6ec25e1000e): Annotation review gallery
1- Add annotation review gallery (see #818 )
2- Add matplotlib and pyvips to setup.py

## [Commit 5](https://github.com/DigitalSlideArchive/HistomicsTK/pull/842/commits/e18ef646a1f03f0f9dae4130372980511cf4d309): Refactor saliency (and other changes)
1- Update setup.py
2- Make sure tests are:
- Actually run by pytest (discovered)
- Don't use unittest module (even disabled ones have been refactored a bit)
- Don't have an ``__init__()`` method so they're not ignored by pytest

3- Exclude pyrtree from coverage since its external code
4- Refactor preprocessing tests (deconvolution, normalization, augmentation) and remove redundant tests
5- Refactor all saliency tests

**All tests pass at this stage. Coverage is ~73%.**

## [Commit 6](https://github.com/DigitalSlideArchive/HistomicsTK/pull/842/commits/4ad127d5da81068d48985e17abd76b90a91099e6): Cherry pick PRs and close stale branch
- Update workflows jupyter
- Cherry-pick PR's #840 and #851 to avoid conflicts.
- Cherry-pick a very old commit (84e3985cb75f20b0094b8db7cc5f0aefa89f28a7) by @cooperlab for the min model function (stale branch)
- Cherry pick PR #765 and resolve conflict

**All tests pass at this stage**

## [Commit 7](https://github.com/DigitalSlideArchive/HistomicsTK/commit/61a8d235c8caee56c7e0445adda18f59d9e5d0a8): Re-enable all tests. 
Refactoring all tests is complete. The only exception is ``test_cli_result.py`` which was tweaked to use the .sha but still needs refactoring for dependence on girder. 

**All tests pass at this stage. Coverage is 76%.**

## [Commit 8](https://github.com/DigitalSlideArchive/HistomicsTK/pull/842/commits/82e4a224aea9dccf7157c4004d4fc8410776de49): Final touches for sphinx docs
- Update docsstrings
- Make sure links are relative to toctree or completely external links (not on a github branch)

**Sphinx docs render without errors/warnings**

### CODE REVIEW

**Comment by kheffah** [comment]
The updated docs (rst files and jupyter) have already been reviewed by @cooperlab . Some have not yet been merged here.

**Comment by kheffah** [comment]
@manthey Thanks! I updated README accordinagly.

**Comment by kheffah** [comment]
@manthey Everything passes, except two errors that seem to be arising from travis:
1- python 2.7 is failing to import some modules
2- For some reason sphinx is failing to import some modules. This is a travis issue because locally, I am able to build the sphinx docs with no warnings.
Do you mind taking a look when you get a chance? Thanks!

**Comment by manthey** [comment]
> 2- For some reason sphinx is failing to import some modules. This is a travis issue because locally, I am able to build the sphinx docs with no warnings.

I see the same warnings about failed imports locally.  I'll look into it.

**Comment by kheffah** [comment]
> > 2- For some reason sphinx is failing to import some modules. This is a travis issue because locally, I am able to build the sphinx docs with no warnings.
> 
> I see the same warnings about failed imports locally. I'll look into it.

Thank you so much, yes I'd appreciate that! I'm assuming I have a different sphinx version then. For some reason, everything runs locally on my computer.

**Comment by kheffah** [comment]
@manthey Thank you for all your help. All checks pass now :)

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -92,45 +92,84 @@ Pull Request Guidelines
...
-   feature to the list in README.rst.
-3. The pull request should work for Python 2.6, 2.7, and for PyPy. Check
+   your new functionality into a function with a docstring, and make sure
+   your feature is reflected in one of the `.rst` files under `./docs/`.
+   If an existing module exists and your feature is appended to it,
+   then you can ignore this step as Sphinx will render it automatically.
+   Otherwise, create a new `.rst` file.
+3. **(Optional)** Create a Jupyter Notebook or a detailed documentation `.rst`
+   under `./docs/examples/` to explain exactly how the feature should be used.
+4. The pull request should work for Python 2.6, 2.7, 3.6, 3.7, and for PyPy. Check
```
Comment (manthey) [inline CONTRIBUTING.rst:104]: As long as we are changing this line, we currently test with 2.7, 3.5, 3.6, and 3.7.  Let's remove 2.6 and PyPy and add 3.5 to the list.

**Comment by manthey** [comment]
@kheffah Assuming you approve of what I've added in PR #864 and that gets merged into this branch, I'd like to then approve this branch and get it merged.  I'm sure based on its scope we'll find something else that needs tweaking, but it is such a great improvement that I'd like to get it merged.

**Comment by kheffah** [comment]
@manthey Fantastic, and thanks so much!! I just reviewed the PR you made and merged it.

**Pull request merged**